Enable REPL to offer to install missing packages if install hooks are provided

IanButterworth opened the pull request:
Just trying out the ideas discussed in https://github.com/JuliaLang/julia/pull/26469 and in https://github.com/JuliaLang/julia/issues/35062 and the suggestion here https://github.com/JuliaLang/julia/pull/26469#issuecomment-597337630 to introduce a specific error type to allow other handlers to decide what to do.

This goes for a more basic `y/n` prompt than the example in https://github.com/JuliaLang/julia/pull/26469, with the default being `n`, which is selected here first.
It also leaves it up to the user to decide if the _active environment_ is appropriate to install into.
A `ctrl-c` acts the same as returning `n` or just pressing `enter`.

Here the default is selected first by just hitting `enter`

[![asciicast](https://asciinema.org/a/kO5Q4sZJp6qfI0a1Gtvo0xArx.svg)](https://asciinema.org/a/kO5Q4sZJp6qfI0a1Gtvo0xArx)

cc. @KristofferC @oxinabox

IanButterworth committed:
REPL: offer to try to install package if not found

KristofferC commented:
Nice! I do think that Pkg itself needs to manage some of this though (and not only REPL). Some comments:

- I think the auto complete should only be done if the package exists in one of the current registries.
- It would be good to be able to enable auto installation of future packages (without requiring a prompt). 
- Also see https://github.com/JuliaLang/julia/pull/26469#issuecomment-523150333
- Regarding the implementation, I think having Pkg register a callback with the REPL (similar to https://github.com/JuliaLang/julia/pull/26469) makes sense because providing the info above will likely require using Pkg internals.

oxinabox commented:
> I think the auto complete should only be done if the package exists in one of the current registries.

Strongly agreed. It would be annoying to get this if you had made a typo in the package name.
and so spent several seconds of it resolving only to get an error message..

> It would be good to be able to enable auto installation of future packages (without requiring a prompt).

I think that can be a seperate PR?
This is useful without it, and the complexity of dealing with how to store persistient settings related to Pkg seems like it would bog this down.
Unless you just mean an eviroment variable that is set externally?

ianshmean commented:
Ok, I've made some changes:

- Switched to a similar hooks method to https://github.com/JuliaLang/julia/pull/26469 but the difference is it hooks into when `ModuleNotFoundError` is thrown during REPL eval, rather than at the `Base.require` level
- Added the appropriate Pkg hook in https://github.com/JuliaLang/Pkg.jl/pull/2307
- Checks for whether the package is available in installed registries. If so, prompts to install, if not, throws the old error directly
- Re-evaluating the ast after the hook is finished, if it returns true, so that the package is loaded at the end. I thought this makes sense given that if the prompted install process completes without error, it would be weird if the original instruction wasn't _successful_

```
julia> using CSVV
ERROR: ModuleNotFoundError: Package CSVV not found in current path:
- Run `import Pkg; Pkg.add("CSVV")` to install the mod package.

Stacktrace:
 [1] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:867

julia> using CSV
 └ Package "CSV" not found, but "CSV" is listed in registry.
   Active project: /Users/ian/Documents/GitHub/Pkg.jl/Project.toml
   Add "CSV" to the active project? (y/n) [n]: y
   Resolving package versions...
Updating `~/Documents/GitHub/Pkg.jl/Project.toml`
  [336ed68f] + CSV v0.8.2
Updating `~/Documents/GitHub/Pkg.jl/Manifest.toml`
  [336ed68f] + CSV v0.8.2
  [e2d170a0] + DataValueInterfaces v1.0.0
  [82899510] + IteratorInterfaceExtensions v1.0.0
  [69de0a69] + Parsers v1.0.15
  [2dfb63ee] + PooledArrays v0.5.3
  [91c51154] + SentinelArrays v1.2.16
  [3783bdb8] + TableTraits v1.0.0
  [bd369af6] + Tables v1.2.2
  Progress [========================================>]  9/9
9 dependencies successfully precompiled in 11 seconds (37 already precompiled)

julia> CSV
CSV
```

DilumAluthge commented:
This is only in interactive usage, right? I.e. in scripts, you'd just throw the error.

ianshmean commented:
Yeah. That's the intention. This is what I've tested
```
ian@Ians-MacBook-Pro-4 julia % ./julia -e "using CSV"
ERROR: ModuleNotFoundError: Package CSV not found in current path:
- Run `import Pkg; Pkg.add("CSV")` to install the mod package.

Stacktrace:
 [1] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:867
```

ianshmean commented:
Just pushed to the corresponding [pkg PR](https://github.com/JuliaLang/Pkg.jl/pull/2307) to improve the info and formatting of the prompt:
<img width="555" alt="Screen Shot 2021-01-02 at 12 33 36 AM" src="https://user-images.githubusercontent.com/1694067/103451461-52061a80-4c92-11eb-8c89-aa4c0564299c.png">

crstnbr commented:
I like this! However, do we need to force the user into a yes/no dialog? These can be rather annoying. What about a more minimal approach, something like

```Julia
julia> using CSV
Package "CSV" not found, but available from a registry.
Install the package with `] add CSV`  or by pressing `CTRL+I`.

julia>
```
Which would install the package upon pressing `CTRL+I` right after such a message has occured.

ianshmean commented:
One option is to make the default `y`, with a short timeout of a few seconds. I opened https://github.com/JuliaLang/julia/pull/39027 if that was desired. 

That would feel less like forcing the user into a decision.

Providing a magic key combo to run the package add is also interesting

StefanKarpinski commented:
The yes/no dialog is what most package managers do here, so it seems pretty standard and easy.

ianshmean commented:
Any further thoughts on this?

In summary I think the flow works well, and it seems like it would allow exotic front ends to do whatever it prefers when the `ModuleNotFoundError` is thrown.

However, I  think there's one (hopefully small) con with this strategy of intercepting errors in the REPL evaluation..
Say the user does these examples, where `Bar` isn't installed
```
julia> using Foo, Bar
```
then `Foo` will be loaded twice, the 2nd time when the entire line is re-evaluated. I guess packages are usually safe to load twice though.

or 
```
julia> x = 1;
julia> x += 1; using Bar
```
that wouldn't be the case with
```
julia> x = 1;
julia> using Bar; x += 1
```

IanButterworth commented:
Just rebased and made this prompt only trigger for lines starting with `using` or `import` to avoid the edge case of re-running code on the same line that precedes the load, shown in the above comment.

Also added "and load" to the end to make it clearer that the package is loaded after install

![image](https://user-images.githubusercontent.com/1694067/111024025-65c4b000-83aa-11eb-9169-0ee7bdff9f8b.png)

simeonschaub commented on stdlib/REPL/src/REPL.jl:
```diff
@@ -774,9 +774,35 @@ find_hist_file() = get(ENV, "JULIA_HISTORY",
+function handle_install_package_hooks(mod, ast, backend)
+    line = string(ast.args[2])
+    # only exercise hooks for lines starting with `using/import` to avoid re-running non-load code
+    if startswith(line, "using") || startswith(line, "import")
```
I am not sure this is the best way to check this, since there could also be leading whitespace or the `using` statement could be further nested but still run first. We could do some analysis on the parsed AST, but I am wondering if in practice just always reevaluating the whole input would really be a problem.

IanButterworth replied:
>but I am wondering if in practice just always reevaluating the whole input would really be a problem.

Yeah, I think you might be right. I've removed the check for using/import

miguelraz commented:
Feature request: if a user does
```julia
using Foo, Bar
```
It will prompt to install all of them. As Stefan says, if users want an in between, they can do it manually.

JeffBezanson commented:
Suggestion from Stefan: After the repl input is parsed, search for `using` and `import` statements in the AST, collect all the names of packages not in the environment, and prompt to install all of them before doing `eval`.

JeffBezanson commented:
The advantage of Stefan's parse-first method is it avoids all issues with when and how many times things are evaluated. Even if the input contains `if false; using X; end`, prompting to install `X` still makes sense, and the conditional `using` still works, since all we've done is modify the project, not the state of the running process.

A "disadvantage" (but depends on your point of view) is that meta-programmed `using` statements will not prompt, but most of us in triage felt that case doesn't matter.

IanButterworth commented:
Now updated along with https://github.com/JuliaLang/Pkg.jl/pull/2307 to @StefanKarpinski's suggestion of checking for all packages that will be loaded in the ast, then handling all packages in one act.

It seems to work quite nicely.

A single package
```
julia> using ImageIO
 │ Package "ImageIO" not found, but "ImageIO" is available from a registry.
 │ Project `~/.julia/environments/v1.7/Project.toml`
 └ Add "ImageIO" to the active project and load? (y/n) [n]: n
ERROR: ArgumentError: Package ImageIO not found in current path:
- Run `import Pkg; Pkg.add("ImageIO")` to install the ImageIO package.

Stacktrace:
 [1] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:873
```

Where all packages are missing, but only two packages are available in a registry
```
julia> using ImageIO, TiffImages, DoesNotExist
 │ Packages ["ImageIO", "TiffImages", "DoesNotExist"] not found, but ["ImageIO", "TiffImages"] are available from a registry.
 │ Project `~/.julia/environments/v1.7/Project.toml`
 └ Add ["ImageIO", "TiffImages"] to the active project and load? (y/n) [n]: n
ERROR: ArgumentError: Package ImageIO not found in current path:
- Run `import Pkg; Pkg.add("ImageIO")` to install the ImageIO package.

Stacktrace:
 [1] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:871
```
Where specific imports are declared
```
julia> using ImageIO: save
 │ Package ["ImageIO"] not found, but ["ImageIO"] is available from a registry.
 │ Project `~/.julia/environments/v1.7/Project.toml`
 └ Add ["ImageIO"] to the active project and load? (y/n) [n]: n
ERROR: ArgumentError: Package ImageIO not found in current path:
- Run `import Pkg; Pkg.add("ImageIO")` to install the ImageIO package.

Stacktrace:
 [1] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:871
```
In guarded code
```
julia> if false if false using ImageIO, TiffImages, DoesNotExist end end
 │ Packages ["ImageIO", "TiffImages", "DoesNotExist"] not found, but ["ImageIO", "TiffImages"] are available from a registry.
 │ Project `~/.julia/environments/v1.7/Project.toml`
 └ Add ["ImageIO", "TiffImages"] to the active project and load? (y/n) [n]: n
```
The first examplle but selecting `y`
```
julia> using ImageIO
 │ Package "ImageIO" not found, but "ImageIO" is available from a registry.
 │ Project `/var/folders/_6/1yf6sj0950vcg4t91m9ltb5w0000gn/T/jl_U2GrdN/Project.toml`
 └ Add "ImageIO" to the active project and load? (y/n) [n]: y
    Updating registry at `~/.julia/registries/General`
    Updating git-repo `https://github.com/JuliaRegistries/General.git`
   Resolving package versions...
   Installed Netpbm ──── v1.0.2
   Installed PNGFiles ── v0.3.7
   Installed ImageCore ─ v0.9.0
    Updating `/private/var/folders/_6/1yf6sj0950vcg4t91m9ltb5w0000gn/T/jl_U2GrdN/Project.toml`
  [82e4d734] + ImageIO v0.5.3
    Updating `/private/var/folders/_6/1yf6sj0950vcg4t91m9ltb5w0000gn/T/jl_U2GrdN/Manifest.toml`
...
Precompiling project...
  40 dependencies successfully precompiled in 48 seconds (1 already precompiled)

julia> ImageIO
ImageIO
```
The second example but selecting `y`
```
julia> using ImageIO, TiffImages, DoesNotExist
 │ Packages ["ImageIO", "TiffImages", "DoesNotExist"] not found, but ["ImageIO", "TiffImages"] are available from a registry.
 │ Project `~/.julia/environments/v1.7/Project.toml`
 └ Add ["ImageIO", "TiffImages"] to the active project and load? (y/n) [n]: y
    Updating registry at `~/.julia/registries/General`
    Updating git-repo `https://github.com/JuliaRegistries/General.git`
   Resolving package versions...
   Installed Netpbm ───── v1.0.1
   Installed TiffImages ─ v0.3.1
    Updating `~/.julia/environments/v1.7/Project.toml`
  [82e4d734] + ImageIO v0.5.3
  [731e570b] + TiffImages v0.3.1
    Updating `~/.julia/environments/v1.7/Manifest.toml`
...
Precompiling project...
...
ERROR: ModuleNotFoundError: Package DoesNotExist not found in current path:
- Run `import Pkg; Pkg.add("DoesNotExist")` to install the mod package.

Stacktrace:
 [1] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:873
```

What it actually looks like on MacOS
![image](https://user-images.githubusercontent.com/1694067/115963682-6e062400-a4ee-11eb-9a3d-6bc48e028275.png)


Doesn't add noticeable overhead to eval:
1.6.1
```
julia> @time @eval 1
  0.000001 seconds
1

julia> @time @eval 1
  0.000003 seconds
1

julia> @time @eval 1
  0.000003 seconds
1

julia> @time @eval 1
  0.000002 seconds
```
This PR
```
julia> @time @eval 1
  0.000002 seconds
1

julia> @time @eval 1
  0.000005 seconds
1

julia> @time @eval 1
  0.000002 seconds
1

julia> @time @eval 1
  0.000001 seconds
1
```

StefanKarpinski commented:
Beautiful!

IanButterworth committed:
tidier code & don't check modules if no hooks available

simeonschaub commented on stdlib/REPL/src/REPL.jl:
```diff
@@ -776,11 +776,46 @@ find_hist_file() = get(ENV, "JULIA_HISTORY",
+            if first(arg.args) isa Symbol # i.e. `Foo`
+                push!(mods, first(arg.args))
+            else # i.e. `Foo: bar`
+                push!(mods, first(first(arg.args).args))
```
I think this might need to be made more robust. I don't think this works for something like `import A.b as c` yet. Does this currently know to properly ignore imports like `using .A`?

IanButterworth replied:
I didn't think about either of those. I'll investigate and add tests

IanButterworth replied:
Fixed the `using .A` case, and the other worked already. Added tests. Are there any other forms to watch out for?

JeffBezanson commented on stdlib/REPL/src/REPL.jl:
```diff
@@ -776,11 +776,46 @@ find_hist_file() = get(ENV, "JULIA_HISTORY",
+
 function eval_with_backend(ast, backend::REPLBackendRef)
+    if !isempty(install_packages_hooks)
+        check_for_missing_packages_and_run_hooks(ast)
```
This should probably run in the backend. It could even use the existing `repl_ast_transforms` hook (not a "transform" per se, but no big deal).

IanButterworth replied:
Ok. Moved to the backend, but given that the `modules_to_be_loaded` parser resides here, it seemed like a misuse to place that on `repl_ast_transforms` given that if IJulia/Pluto etc. wanted to do a different install hook they'd probably want to go after the modules have been identified by `modules_to_be_loaded` so keeping `install_packages_hooks` would be necessary anyway. These use models are a bit imaginary though.. my thinking might be off

IanButterworth committed:
exclude dotted local imports

IanButterworth committed:
add tests for loaded modules parser

IanButterworth committed:
run in backend
